# NYC Storms & 311 Flooding Service Requests Part 1: Data Preparation 

**[part 1](https://dvignoles.github.io/postgis-book/dv-finalproject-01.html)**

[part 2](https://dvignoles.github.io/postgis-book/dv-finalproject-2.html)

[part 3](https://dvignoles.github.io/postgis-book/dv-finalproject-3.html)

The goal of this project is to:

(A) Use the local environmental record to identify storms in NYC from 2010-present

(B) Use flooding service requests made to 311 to identify storms in NYC from 2010-present

(C) Use the spatial distribution of flooding service requests made to 311 for each identified storm as proxy for flooding susceptibility due to storms. 

This section covers loading our data sources into PostGIS and creating needed summary tables. 


In [1]:
import requests
import json
import pandas as pd
import geopandas as gpd
from io import StringIO
from pathlib import Path
from sqlalchemy import create_engine

If database not created run "`make create-storms-db`". See `makefile` in project repo for details.

In [2]:
engine_str = "postgresql+psycopg2://docker:docker@0.0.0.0:25432/storms"
engine = create_engine(engine_str)

In [3]:
%load_ext sql
%sql $engine.url

'Connected: docker@storms'

In [4]:
%sql CREATE EXTENSION IF NOT EXISTS postgis;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.


[]

## Site Metadata

Several of the data products are time series "panel" data at georeferenced sites. To maintain database normality, let's store the site information in a separate table.

In [7]:
%%sql 
CREATE TABLE IF NOT EXISTS sites (
    id serial UNIQUE,
    source varchar(20),
    source_site_id varchar(20),
    site_name varchar(50),
    geom geometry(POINT)
);

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.


[]

## Daily Streamflow


In [8]:
%%sql tables <<
SELECT * FROM pg_catalog.pg_tables WHERE schemaname='public' AND tableowner='docker';

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
20 rows affected.
Returning data to local variable tables


In [12]:
%%sql
INSERT INTO sites (source, source_site_id, site_name, geom)
VALUES ('USGS', '01302020', 'BRONX RIVER AT NY BOTANICAL GARDEN AT BRONX NY', ST_Transform(ST_Point(-73.87438889, 40.86230556, 4269), 4263))
ON CONFLICT DO NOTHING

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
0 rows affected.


[]

In [13]:
usgs_csv = Path('../data/usgs_bronx.csv')

if not usgs_csv.exists():
    resp = requests.get('https://nwis.waterdata.usgs.gov/usa/nwis/uv/?cb_00060=on&cb_00065=on&format=rdb&site_no=01302020&legacy=1&period=&begin_date=2010-01-01&end_date=2022-12-11')
    usgs_raw = resp.content.decode('utf-8')
    usgs_df = pd.read_csv(StringIO(usgs_raw), delimiter='\t', skiprows=list(range(0,30)) + [31,])
    usgs_df.to_csv('../data/usgs_bronx.csv', index=False)
else:
    usgs_df = pd.read_csv(usgs_csv)

In [14]:
usgs_df.loc[:, 'datetime'] = pd.to_datetime(usgs_df.datetime, format='%Y-%m-%d %H:%M')
usgs_df.set_index('datetime', inplace=True)
usgs_df.rename(columns={'106409_00065': 'gage_height_ft', '106410_00060': 'discharge_cfs'}, inplace=True)
usgs_df.loc[:, 'site_id'] = 1
usgs_df = usgs_df[['site_id', 'gage_height_ft', 'discharge_cfs']]

/tmp/ipykernel_2070666/300101538.py:1: FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  usgs_df.loc[:, 'datetime'] = pd.to_datetime(usgs_df.datetime, format='%Y-%m-%d %H:%M')


In [15]:
usgs_df.head()

site_id  gage_height_ft  discharge_cfs
datetime                                                   
2010-01-01 00:00:00        1            0.97           94.9
2010-01-01 00:15:00        1            0.97           94.9
2010-01-01 00:30:00        1            0.97           94.9
2010-01-01 00:45:00        1            0.97           94.9
2010-01-01 01:00:00        1            0.97           94.9

In [16]:
%%sql tables <<
SELECT * FROM pg_catalog.pg_tables WHERE schemaname='public' AND tableowner='docker';

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
20 rows affected.
Returning data to local variable tables


In [17]:
if "usgs_bronx_gage" not in tables.DataFrame().tablename.tolist():
    with engine.connect() as con:
        usgs_df.to_sql('usgs_bronx_gage', con, if_exists='replace')

We are only interested in values at the daily granularity. The USGS values are reported with a 15 minute temporal resolution. 

In [18]:
%%sql 
-- create daily table
CREATE TABLE IF NOT EXISTS usgs_bronx_gage_daily AS 

    SELECT 
        DATE_TRUNC('day', datetime)::date as date, 
        site_id, 
        MAX(gage_height_ft) as gage_height_ft_max, 
        AVG(discharge_cfs) as discharge_cfs
    FROM usgs_bronx_gage
    GROUP BY DATE_TRUNC('day', datetime), site_id
    ORDER BY date;

-- foreign key
ALTER TABLE usgs_bronx_gage_daily DROP CONSTRAINT IF EXISTS fk_site_gage;
ALTER TABLE usgs_bronx_gage_daily
    ADD CONSTRAINT fk_site_gage FOREIGN KEY (site_id)
        REFERENCES sites(id);

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
Done.
Done.


[]

In [19]:
%sql SELECT * FROM usgs_bronx_gage_daily LIMIT 5;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
5 rows affected.


[(datetime.date(2010, 1, 1), 1, 0.97, 89.38125000000002),
 (datetime.date(2010, 1, 2), 1, 0.94, 77.6625),
 (datetime.date(2010, 1, 3), 1, 0.88, 67.86458333333321),
 (datetime.date(2010, 1, 4), 1, 0.85, 60.871875000000074),
 (datetime.date(2010, 1, 5), 1, 0.83, 58.69687499999987)]

We will be comparing daily values to historical monthly long term means to detect possible storm events. We will also calculate the 25th quartile, 75th quartiles, and the statistical maximums for each month.

In [20]:
%%sql
CREATE TABLE IF NOT EXISTS usgs_bronx_gage_mlt AS
SELECT 
    EXTRACT('MONTH' FROM date) as month, 
    site_id, 
    AVG(gage_height_ft_max) as gage_height_ft_max_avg, 
    PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY gage_height_ft_max) AS gage_height_ft_max_iqr25,
    PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY gage_height_ft_max) AS gage_height_ft_max_iqr75, 
    AVG(discharge_cfs) as discharge_cfs_avg,
    PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY discharge_cfs) AS discharge_cfs_iqr25,
    PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY discharge_cfs) AS discharge_cfs_iqr75
FROM usgs_bronx_gage_daily
GROUP BY EXTRACT('MONTH' FROM date) , site_id
ORDER BY month;

ALTER TABLE usgs_bronx_gage_mlt
    ADD COLUMN IF NOT EXISTS gage_height_ft_max_iqr numeric,
    ADD COLUMN IF NOT EXISTS gage_height_ft_max_iqrmax numeric,
    ADD COLUMN IF NOT EXISTS discharge_cfs_iqr numeric,
    ADD COLUMN IF NOT EXISTS discharge_cfs_iqrmax numeric;

UPDATE usgs_bronx_gage_mlt
    SET gage_height_ft_max_iqr = gage_height_ft_max_iqr75 - gage_height_ft_max_iqr25,
    discharge_cfs_iqr = discharge_cfs_iqr75 - discharge_cfs_iqr25;

-- 75th quartile + 1.5 x Interquartile Range
UPDATE usgs_bronx_gage_mlt
    SET gage_height_ft_max_iqrmax = gage_height_ft_max_iqr75 + (1.5 * gage_height_ft_max_iqr),
    discharge_cfs_iqrmax = discharge_cfs_iqr75 + (1.5 * discharge_cfs_iqr);

-- foreign key
ALTER TABLE usgs_bronx_gage_mlt DROP CONSTRAINT IF EXISTS fk_site_gage_mlt;
ALTER TABLE usgs_bronx_gage_mlt
    ADD CONSTRAINT fk_site_gage_mlt FOREIGN KEY (site_id)
        REFERENCES sites(id);
SELECT * FROM usgs_bronx_gage_mlt;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
Done.
12 rows affected.
12 rows affected.
Done.
Done.
12 rows affected.


[(Decimal('1'), 1, 0.9028287841191059, 0.7, 0.93, 73.12444870177197, 38.24947916666673, 79.12187500000006, Decimal('0.23'), Decimal('1.275'), Decimal('40.8723958333333'), Decimal('140.43046875')),
 (Decimal('2'), 1, 0.9465667574931881, 0.73, 1.01, 81.98060629742612, 43.802083333333314, 89.00624999999995, Decimal('0.28'), Decimal('1.43'), Decimal('45.2041666666666'), Decimal('156.8125')),
 (Decimal('3'), 1, 1.0612158808933003, 0.79, 1.13, 116.3898950803755, 56.221875000000026, 116.05208333333334, Decimal('0.34'), Decimal('1.64'), Decimal('59.8302083333333'), Decimal('205.797395833333')),
 (Decimal('4'), 1, 1.0611794871794873, 0.77, 1.15, 105.55898100573297, 51.30442708333338, 105.87447916666665, Decimal('0.38'), Decimal('1.72'), Decimal('54.5700520833333'), Decimal('187.729557291667')),
 (Decimal('5'), 1, 1.024689826302731, 0.73, 1.12, 86.57926350512383, 39.74687500000002, 92.8229166666666, Decimal('0.39'), Decimal('1.705'), Decimal('53.0760416666666'), Decimal('172.436979166667')),
 (Decimal('6'), 1, 0.9140000000000004, 0.66, 0.9875, 65.52605934270572, 30.99947916666668, 65.25963541666677, Decimal('0.3275'), Decimal('1.47875'), Decimal('34.2601562500001'), Decimal('116.649869791667')),
 (Decimal('7'), 1, 0.8712219451371576, 0.58, 0.98, 56.013134914141624, 23.018750000000047, 55.81875000000002, Decimal('0.4'), Decimal('1.58'), Decimal('32.8'), Decimal('105.01875')),
 (Decimal('8'), 1, 0.8190570719602984, 0.53, 0.85, 56.31492750176646, 18.4703125, 46.270312499999974, Decimal('0.32'), Decimal('1.33'), Decimal('27.8'), Decimal('87.9703125')),
 (Decimal('9'), 1, 0.8403589743589742, 0.53, 0.8474999999999999, 65.21470956651214, 17.015625, 51.326562499999966, Decimal('0.3175'), Decimal('1.32375'), Decimal('34.3109375'), Decimal('102.79296875')),
 (Decimal('10'), 1, 0.8362779156327539, 0.57, 0.885, 57.25864653667988, 19.55052083333335, 57.56927083333332, Decimal('0.315'), Decimal('1.3575'), Decimal('38.01875'), Decimal('114.597395833333')),
 (Decimal('11'), 1, 0.8609487179487186, 0.61, 0.88, 61.269140322762134, 28.62656250000005, 67.10286458333321, Decimal('0.27'), Decimal('1.285'), Decimal('38.4763020833332'), Decimal('124.817317708333')),
 (Decimal('12'), 1, 0.951305483028721, 0.655, 1.03, 80.33195802150934, 31.07135416666672, 88.98020833333332, Decimal('0.375'), Decimal('1.5925'), Decimal('57.9088541666666'), Decimal('175.843489583333'))]

Finally, let's create a table of dates where the discharge or gage height exceeds its statistical maximum value.

In [21]:
%%sql
DROP TABLE IF EXISTS usgs_bronx_gage_daily_high_events;
CREATE TABLE usgs_bronx_gage_daily_high_events AS
WITH highs AS (
    SELECT 
    dts.site_id,
    dts.date,
    discharge_cfs,
    gage_height_ft_max,
    CASE
        WHEN discharge_cfs > discharge_cfs_iqr75 THEN 1
        ELSE 0
    END discharge_cfs_gt75,
    discharge_cfs - discharge_cfs_iqr75 as discharge_cfs_diff75,
    CASE
        WHEN discharge_cfs > discharge_cfs_iqrmax THEN 1
        ELSE 0
    END discharge_cfs_gtmax,
    discharge_cfs - discharge_cfs_iqrmax as discharge_cfs_diffmax,
    CASE
        WHEN gage_height_ft_max >= gage_height_ft_max_iqr75 THEN 1
        ELSE 0
    END gage_height_ft_max_gt75,
    gage_height_ft_max - gage_height_ft_max_iqr75 as gage_height_ft_max_diff75,
    CASE
        WHEN gage_height_ft_max >= gage_height_ft_max_iqrmax THEN 1
        ELSE 0
    END gage_height_ft_max_gtmax,
    gage_height_ft_max - gage_height_ft_max_iqrmax as gage_height_ft_max_diffmax
    FROM
    usgs_bronx_gage_daily dts
    INNER JOIN usgs_bronx_gage_mlt mlt ON EXTRACT('MONTH' FROM dts.date) = mlt.month
    ORDER BY dts.date
)
SELECT 
    highs.*
FROM highs
WHERE 
    highs.discharge_cfs_gtmax = 1
        OR
    highs.gage_height_ft_max_gtmax = 1;

SELECT * FROM usgs_bronx_gage_daily_high_events
ORDER BY discharge_cfs_diff75 DESC
LIMIT 10;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
562 rows affected.
10 rows affected.


[(1, datetime.date(2011, 8, 28), 2202.0208333333335, 5.18, 1, 2155.7505208333337, 1, 2114.0505208333334, 1, 4.33, 1, 3.8499999999999996),
 (1, datetime.date(2021, 9, 2), 2134.28125, 5.59, 1, 2082.9546875, 1, 2031.48828125, 1, 4.7425, 1, 4.266249999999999),
 (1, datetime.date(2011, 3, 11), 1680.3125, 4.34, 1, 1564.2604166666667, 1, 1474.515104166667, 1, 3.21, 1, 2.7),
 (1, datetime.date(2011, 9, 8), 1300.4270833333333, 3.71, 1, 1249.1005208333333, 1, 1197.6341145833333, 1, 2.8625, 1, 2.38625),
 (1, datetime.date(2011, 4, 17), 1341.7395833333333, 4.06, 1, 1235.8651041666667, 1, 1154.0100260416662, 1, 2.9099999999999997, 1, 2.34),
 (1, datetime.date(2011, 3, 7), 1190.9895833333333, 3.58, 1, 1074.9375, 1, 985.1921875000003, 1, 2.45, 1, 1.9400000000000002),
 (1, datetime.date(2018, 4, 16), 1179.4895833333333, 4.24, 1, 1073.6151041666667, 1, 991.7600260416663, 1, 3.0900000000000003, 1, 2.5200000000000005),
 (1, datetime.date(2010, 3, 30), 1117.28125, 3.61, 1, 1001.2291666666666, 1, 911.483854166667, 1, 2.48, 1, 1.97),
 (1, datetime.date(2014, 5, 1), 1074.4895833333333, 3.68, 1, 981.6666666666666, 1, 902.0526041666662, 1, 2.56, 1, 1.975),
 (1, datetime.date(2010, 3, 14), 1022.5729166666666, 3.26, 1, 906.5208333333333, 1, 816.7755208333336, 1, 2.13, 1, 1.6199999999999999)]

The top two entries correspond to the dates of hurricanes Irene and Ida hitting NYC. 

In [22]:
%%sql 
SELECT COUNT(*) 
FROM usgs_bronx_gage_daily_high_events
WHERE discharge_cfs_diffmax = 1
OR gage_height_ft_max_gtmax  = 1;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
1 rows affected.


[(471,)]

There were 471 days with values exceeding a statistical maximum for the gage. 

## Daily Tidal Highs
We are interested in the daily high water level values

In [24]:
%%sql
INSERT INTO sites (source, source_site_id, site_name, geom)
VALUES ('NOAA', '8518750', 'The Battery, NY', ST_Point(-74.0142, 40.7006, 4263))
ON CONFLICT DO NOTHING;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
0 rows affected.


[]

In [25]:
%%sql tables <<
SELECT * FROM pg_catalog.pg_tables WHERE schemaname='public' AND tableowner='docker';

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
20 rows affected.
Returning data to local variable tables


In [26]:
if 'tidal_battery_highs_daily' not in tables.DataFrame().tablename.tolist():
    with engine.connect() as con:
        for year in range(2010,2023):
            mode = 'replace' if year == 2010 else 'append'
            url = f'https://api.tidesandcurrents.noaa.gov/api/prod/datagetter?begin_date={year}0101&end_date={year}1211&station=8518750&product=high_low&datum=STND&time_zone=lst&units=english&format=json'
            resp = requests.get(url)
            j = json.loads(resp.content.decode('utf-8'))
            df = pd.DataFrame.from_records(j['data'])
            df.loc[:, 'datetime'] = pd.to_datetime(df.t)
            df.loc[:, 'site_id'] = 2
            df = df[df.ty.str.contains('H')][['site_id', 'datetime', 'v',]].rename(columns={'v': 'waterlevel_stnd_ft',})
            df.loc[:, 'waterlevel_stnd_ft'] = df.waterlevel_stnd_ft.astype(pd.Float64Dtype())
            df.set_index('datetime', inplace=True)
            df.to_sql('tidal_battery_highs_daily', con, if_exists=mode)

In [27]:
%%sql
ALTER TABLE tidal_battery_highs_daily DROP CONSTRAINT IF EXISTS fk_site_tides;
ALTER TABLE tidal_battery_highs_daily
    ADD CONSTRAINT fk_site_tides FOREIGN KEY (site_id)
        REFERENCES sites(id);

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
Done.


[]

In [28]:
%sql SELECT * FROM tidal_battery_highs_daily LIMIT 5;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
5 rows affected.


[(datetime.datetime(2010, 1, 1, 8, 12), 2, 9.869),
 (datetime.datetime(2010, 1, 1, 20, 42), 2, 8.996),
 (datetime.datetime(2010, 1, 2, 9, 18), 2, 9.865),
 (datetime.datetime(2010, 1, 2, 21, 36), 2, 7.717),
 (datetime.datetime(2010, 1, 3, 10, 6), 2, 7.631)]

Let's calculate the monthly IQR values using the same process as with the Bronx streamflow gage.

In [29]:

%%sql
CREATE TABLE IF NOT EXISTS tidal_battery_highs_mlt AS
SELECT 
    EXTRACT('MONTH' FROM datetime) as month, 
    site_id, 
    AVG(waterlevel_stnd_ft) as waterlevel_stnd_ft_avg, 
    PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY waterlevel_stnd_ft) AS waterlevel_stnd_ft_iqr25,
    PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY waterlevel_stnd_ft) AS waterlevel_stnd_ft_iqr75
FROM tidal_battery_highs_daily
GROUP BY EXTRACT('MONTH' FROM datetime) , site_id
ORDER BY month;

ALTER TABLE tidal_battery_highs_mlt
    ADD COLUMN IF NOT EXISTS waterlevel_stnd_ft_iqr numeric,
    ADD COLUMN IF NOT EXISTS waterlevel_stnd_ft_iqrmax numeric;

UPDATE tidal_battery_highs_mlt
    SET waterlevel_stnd_ft_iqr = waterlevel_stnd_ft_iqr75 - waterlevel_stnd_ft_iqr25;

-- 75th quartile + 1.5 x Interquartile Range
UPDATE tidal_battery_highs_mlt
    SET waterlevel_stnd_ft_iqrmax = waterlevel_stnd_ft_iqr75 + (1.5 * waterlevel_stnd_ft_iqr);

-- foreign key
ALTER TABLE tidal_battery_highs_mlt DROP CONSTRAINT IF EXISTS fk_site_tidal_mlt;
ALTER TABLE tidal_battery_highs_mlt
    ADD CONSTRAINT fk_site_tidal_mlt FOREIGN KEY (site_id)
        REFERENCES sites(id);

SELECT * FROM tidal_battery_highs_mlt;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
Done.
12 rows affected.
12 rows affected.
Done.
Done.
12 rows affected.


[(Decimal('1'), 2, 8.094671373555846, 7.464, 8.7185, Decimal('1.2545'), Decimal('10.60025')),
 (Decimal('2'), 2, 8.03254442877292, 7.402, 8.691, Decimal('1.289'), Decimal('10.6245')),
 (Decimal('3'), 2, 8.266475703324808, 7.657, 8.91, Decimal('1.253'), Decimal('10.7895')),
 (Decimal('4'), 2, 8.424626826029217, 7.894, 8.894, Decimal('1'), Decimal('10.394')),
 (Decimal('5'), 2, 8.486006426735218, 8.012, 8.901, Decimal('0.888999999999999'), Decimal('10.2345')),
 (Decimal('6'), 2, 8.605956233421747, 8.186, 8.98, Decimal('0.794'), Decimal('10.171')),
 (Decimal('7'), 2, 8.599311538461539, 8.16825, 9.0045, Decimal('0.83625'), Decimal('10.258875')),
 (Decimal('8'), 2, 8.666285347043699, 8.241999999999999, 9.098, Decimal('0.856000000000002'), Decimal('10.382')),
 (Decimal('9'), 2, 8.749167330677286, 8.268, 9.186, Decimal('0.917999999999999'), Decimal('10.563')),
 (Decimal('10'), 2, 8.740281853281854, 8.205, 9.265, Decimal('1.06'), Decimal('10.855')),
 (Decimal('11'), 2, 8.371181034482765, 7.85025, 8.88875, Decimal('1.0385'), Decimal('10.4465')),
 (Decimal('12'), 2, 8.34051968503937, 7.71025, 8.89925, Decimal('1.189'), Decimal('10.68275'))]

To do the tide cycle, there are usually two peak water level values for a given day. We will create a view with the maximum peak for each date.

In [30]:
%%sql
CREATE OR REPLACE VIEW tidal_battery_highs_daily_max AS
    SELECT 
    dts.site_id,
    dts.datetime::date as date,
    MAX(waterlevel_stnd_ft) AS waterlevel_stnd_ft
    FROM tidal_battery_highs_daily dts
    GROUP BY dts.datetime::date, site_id
    ORDER BY date;

SELECT * FROM tidal_battery_highs_daily_max LIMIT 5;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
5 rows affected.


[(2, datetime.date(2010, 1, 1), 9.869),
 (2, datetime.date(2010, 1, 2), 9.865),
 (2, datetime.date(2010, 1, 3), 7.631),
 (2, datetime.date(2010, 1, 4), 8.944),
 (2, datetime.date(2010, 1, 5), 8.53)]

In [31]:
%%sql
CREATE TABLE IF NOT EXISTS tidal_battery_highs_events AS
WITH highs AS (
    SELECT 
    daily.site_id,
    daily.date,
    daily.waterlevel_stnd_ft,
    CASE
        WHEN waterlevel_stnd_ft > waterlevel_stnd_ft_iqr75 THEN 1
        ELSE 0
    END waterlevel_stnd_ft_gt75,
    waterlevel_stnd_ft - waterlevel_stnd_ft_iqr75 as waterlevel_stnd_ft_diff75,
    CASE
        WHEN waterlevel_stnd_ft > waterlevel_stnd_ft_iqrmax THEN 1
        ELSE 0
    END waterlevel_stnd_ft_gtmax,
    waterlevel_stnd_ft - waterlevel_stnd_ft_iqrmax as waterlevel_stnd_ft_diffmax
    FROM tidal_battery_highs_daily_max daily
    INNER JOIN tidal_battery_highs_mlt mlt ON EXTRACT('MONTH' FROM daily.date) = mlt.month
    ORDER BY daily.date
)
SELECT 
    highs.*
FROM highs
WHERE 
    highs.waterlevel_stnd_ft_gtmax = 1;


SELECT * FROM tidal_battery_highs_events
ORDER BY waterlevel_stnd_ft_diff75 DESC
LIMIT 10;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
10 rows affected.


[(2, datetime.date(2012, 10, 29), 17.329, 1, 8.064, 1, 6.474),
 (2, datetime.date(2011, 8, 28), 12.795, 1, 3.696999999999999, 1, 2.4130000000000003),
 (2, datetime.date(2010, 3, 13), 12.09, 1, 3.1799999999999997, 1, 1.3004999999999995),
 (2, datetime.date(2016, 2, 9), 11.24, 1, 2.5489999999999995, 1, 0.6155000000000008),
 (2, datetime.date(2011, 4, 16), 11.385, 1, 2.4909999999999997, 1, 0.9909999999999997),
 (2, datetime.date(2016, 1, 10), 11.165, 1, 2.4464999999999986, 1, 0.5647499999999983),
 (2, datetime.date(2016, 2, 8), 11.102, 1, 2.4109999999999996, 1, 0.4775000000000009),
 (2, datetime.date(2018, 10, 27), 11.66, 1, 2.3949999999999996, 1, 0.8049999999999997),
 (2, datetime.date(2021, 2, 1), 11.076, 1, 2.385, 1, 0.4515000000000011),
 (2, datetime.date(2018, 3, 3), 11.188, 1, 2.2780000000000005, 1, 0.3985000000000003)]

Our highest highs correspond to major storm events in NYC as expected.

In [32]:
%%sql
SELECT COUNT(*) 
FROM tidal_battery_highs_events
WHERE waterlevel_stnd_ft_gtmax = 1;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
1 rows affected.


[(56,)]

## Precipitation Daily Summaries

In [33]:
precip_csv = Path('../data/precipitation_nyc.csv')
precip_df = pd.read_csv(precip_csv)

In [34]:
stations = precip_df.groupby('STATION').first().reset_index()[['STATION', 'NAME', 'LATITUDE', 'LONGITUDE']]
site_id = 3

with engine.connect() as con:
    for s in stations.iterrows():
        station = s[1]
        sql = ("INSERT INTO sites (id, source, source_site_id, site_name, geom) "
        f"VALUES ({site_id}, 'NCEI', '{station.STATION}', '{station.NAME}', "
        f"ST_Point({station.LONGITUDE}, {station.LATITUDE}, 4263)) ON CONFLICT DO NOTHING;")
        
        precip_df.loc[precip_df.STATION == station.STATION, 'site_id'] = int(site_id)
        site_id += 1
        con.execute(sql)

In [35]:
precip_df = precip_df[['site_id', 'DATE', 'PRCP']]
precip_df.loc[:, 'site_id'] = precip_df.site_id.astype(int)
precip_df.loc[:, 'DATE'] = pd.to_datetime(precip_df.DATE, format='%Y-%m-%d')
precip_df.rename(columns={'PRCP': 'prcp', 'DATE':'date'}, inplace=True)
precip_df.set_index(['site_id', 'date'], inplace=True)
precip_df.head()

/tmp/ipykernel_2070666/2531588765.py:2: FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  precip_df.loc[:, 'site_id'] = precip_df.site_id.astype(int)
/tmp/ipykernel_2070666/2531588765.py:3: FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  precip_df.loc[:, 'DATE'] = pd.to_datetime(precip_df.DATE, format='%Y-%m-%d')


prcp
site_id date            
6       2010-01-01  0.03
        2010-01-02  0.02
        2010-01-03  0.00
        2010-01-04  0.00
        2010-01-05  0.00

In [36]:
%%sql tables <<
SELECT * FROM pg_catalog.pg_tables WHERE schemaname='public' AND tableowner='docker';

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
20 rows affected.
Returning data to local variable tables


In [37]:
if 'ncei_precip_daily' not in tables.DataFrame().tablename.tolist():
    with engine.connect() as con:
        precip_df.to_sql('ncei_precip_daily', con, if_exists='replace')

In [38]:
%sql SELECT * FROM ncei_precip_daily LIMIT 5;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
5 rows affected.


[(3, datetime.datetime(2022, 5, 30, 0, 0), None),
 (6, datetime.datetime(2010, 1, 1, 0, 0), 0.03),
 (6, datetime.datetime(2010, 1, 2, 0, 0), 0.02),
 (6, datetime.datetime(2010, 1, 3, 0, 0), 0.0),
 (6, datetime.datetime(2010, 1, 4, 0, 0), 0.0)]

In [39]:
%%sql
ALTER TABLE ncei_precip_daily DROP CONSTRAINT IF EXISTS fk_site_tides;
ALTER TABLE ncei_precip_daily
    ADD CONSTRAINT fk_site_tides FOREIGN KEY (site_id)
        REFERENCES sites(id);

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
Done.


[]

Lets create a view for each station for later convenience.

In [40]:
%%sql
-- staten island
CREATE OR REPLACE VIEW ncei_precip_daily_si AS
SELECT * FROM ncei_precip_daily WHERE site_id = 3;

-- laguardia airport
CREATE OR REPLACE VIEW ncei_precip_daily_lag AS
SELECT * FROM ncei_precip_daily WHERE site_id = 4;

-- newark airport
CREATE OR REPLACE VIEW ncei_precip_daily_nwk AS
SELECT * FROM ncei_precip_daily WHERE site_id = 5;

-- central park
CREATE OR REPLACE VIEW ncei_precip_daily_cp AS
SELECT * FROM ncei_precip_daily WHERE site_id = 6;

-- JFK airport 
CREATE OR REPLACE VIEW ncei_precip_daily_jfk AS
SELECT * FROM ncei_precip_daily WHERE site_id = 7;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
Done.
Done.
Done.
Done.


[]

In the case of precipitation, many of the daily sums are zero. The 75th quartile is quite low, so we will use the 95th and 99th percentile values as our markers for "significant precipitation" 

In [41]:

%%sql
CREATE TABLE IF NOT EXISTS ncei_precip_mlt AS
SELECT 
    EXTRACT('MONTH' FROM date) as month, 
    site_id, 
    AVG(prcp) as prcp_avg, 
    PERCENTILE_CONT(0.95) WITHIN GROUP (ORDER BY prcp) AS prcp_95,
    PERCENTILE_CONT(0.99) WITHIN GROUP (ORDER BY prcp) AS prcp_99
FROM ncei_precip_daily
GROUP BY EXTRACT('MONTH' FROM date) , site_id
ORDER BY month;

-- foreign key
ALTER TABLE ncei_precip_mlt DROP CONSTRAINT IF EXISTS fk_site_tidal_mlt;
ALTER TABLE ncei_precip_mlt
    ADD CONSTRAINT fk_site_tidal_mlt FOREIGN KEY (site_id)
        REFERENCES sites(id);

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
Done.
Done.


[]

In [42]:
%sql SELECT * FROM ncei_precip_mlt LIMIT 5;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
5 rows affected.


[(Decimal('1'), 3, 0.11689349112426037, 0.6774999999999989, 1.7777999999999998),
 (Decimal('1'), 4, 0.09960297766749383, 0.6499999999999978, 1.2180000000000017),
 (Decimal('1'), 5, 0.10267990074441688, 0.6289999999999998, 1.299000000000001),
 (Decimal('1'), 6, 0.11054590570719602, 0.7079999999999995, 1.3292000000000008),
 (Decimal('1'), 7, 0.10367245657568241, 0.6279999999999996, 1.279000000000001)]

In [43]:
%%sql 
CREATE TABLE IF NOT EXISTS ncei_precip_daily_events AS
WITH highs as(
    SELECT dts.site_id, dts.date,
    prcp,
    CASE
        WHEN prcp > prcp_99 THEN 1
        ELSE 0
    END prcp_gt99,
    prcp - prcp_99 as prcp_diff99
    FROM ncei_precip_daily dts
    JOIN ncei_precip_mlt mlt
    ON dts.site_id = mlt.site_id
    AND EXTRACT('MONTH' FROM dts.date) = mlt.month
    ORDER BY dts.date
)
SELECT 
    highs.*
FROM highs
WHERE 
    highs.prcp_gt99 = 1;

-- demo
SELECT * FROM ncei_precip_daily_events
ORDER BY prcp_diff99 DESC
LIMIT 10;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
10 rows affected.


[(5, datetime.datetime(2021, 9, 1, 0, 0), 8.41, 1, 6.9789),
 (7, datetime.datetime(2011, 8, 14, 0, 0), 7.8, 1, 5.461599999999999),
 (6, datetime.datetime(2021, 9, 1, 0, 0), 7.13, 1, 5.1789),
 (4, datetime.datetime(2021, 9, 1, 0, 0), 6.8, 1, 5.0769999999999955),
 (3, datetime.datetime(2021, 9, 2, 0, 0), 6.78, 1, 4.911800000000001),
 (3, datetime.datetime(2011, 8, 28, 0, 0), 7.12, 1, 4.7378000000000045),
 (5, datetime.datetime(2011, 8, 14, 0, 0), 6.4, 1, 4.712999999999997),
 (4, datetime.datetime(2011, 8, 14, 0, 0), 6.6, 1, 4.659199999999991),
 (4, datetime.datetime(2014, 4, 30, 0, 0), 5.26, 1, 3.646799999999998),
 (5, datetime.datetime(2011, 8, 28, 0, 0), 5.22, 1, 3.532999999999997)]

In [44]:
%sql SELECT COUNT(*) FROM ncei_precip_daily_events;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
1 rows affected.


[(260,)]

There were 260 99th% precipitation events among the 4 stations from 2010-2022. 

## 311 Service Requests containing "flooding"

In [45]:
flood_csv = Path('../data/flooding_311.csv')
flood_df = pd.read_csv(flood_csv)

/tmp/ipykernel_2070666/4021587811.py:2: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.
  flood_df = pd.read_csv(flood_csv)


In [46]:
%%sql tables <<
SELECT * FROM pg_catalog.pg_tables WHERE schemaname='public' AND tableowner='docker';

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
20 rows affected.
Returning data to local variable tables


In [47]:
def _rename(col):
    col = col.strip()
    col = col.replace(' ', '_')
    col = col.lower()
    col = col.replace('(','')
    col = col.replace(')','')
    return col

if 'nyc_311_flooding' not in tables.DataFrame().tablename.tolist():
    flood_df.rename(mapper=_rename, axis=1, inplace=True)
    flood_df.loc[:, 'created_date'] = pd.to_datetime(flood_df.created_date,format='%m/%d/%Y %I:%M:%S %p')
    flood_df.loc[:, 'closed_date'] = pd.to_datetime(flood_df.closed_date,format='%m/%d/%Y %I:%M:%S %p')
    flood_df.head()

In [48]:
if 'nyc_311_flooding' not in tables.DataFrame().tablename.tolist():
    with engine.connect() as con:
        flood_df.to_sql('nyc_311_flooding', con, if_exists='replace')

In [49]:
%%sql
-- remove non-georeferenced complaints
DELETE FROM nyc_311_flooding
WHERE longitude IS NULL
OR latitude is NULL;

-- add geometry
ALTER TABLE nyc_311_flooding
ADD COLUMN IF NOT EXISTS geom geometry(POINT);

UPDATE nyc_311_flooding
SET geom = ST_Point(longitude, latitude, 4326);

-- add geography
ALTER TABLE nyc_311_flooding
ADD COLUMN IF NOT EXISTS geog geography(POINT);

UPDATE nyc_311_flooding
SET geog = geom::geography;


-- long island state plane
ALTER TABLE nyc_311_flooding
ADD COLUMN IF NOT EXISTS geom_2263 geometry(POINT);

UPDATE nyc_311_flooding
SET geom_2263 = ST_Transform(geom, 2263);


-- add spatial index
CREATE INDEX IF NOT EXISTS nyc_flooding_311_geom_idx on nyc_311_flooding USING gist(geom);
CREATE INDEX IF NOT EXISTS nyc_flooding_311_geom_2263_idx on nyc_311_flooding USING gist(geom_2263);
CREATE INDEX IF NOT EXISTS nyc_flooding_311_geog_idx on nyc_311_flooding USING gist(geog);

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
0 rows affected.
Done.
136108 rows affected.
Done.
136108 rows affected.
Done.
136108 rows affected.
Done.
Done.
Done.


[]

In [50]:
%sql SELECT DISTINCT descriptor FROM nyc_311_flooding;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
4 rows affected.


[('RAIN GARDEN FLOODING (SRGFLD)',),
 ('Street Flooding (SJ)',),
 ('Highway Flooding (SH)',),
 ('Catch Basin Clogged/Flooding (Use Comments) (SC)',)]

In [51]:
%%sql
CREATE OR REPLACE VIEW nyc_311_flooding_daily_complaints AS
    SELECT 
    created_date::date as date,
    COUNT(created_date) as num_complaints
    FROM nyc_311_flooding
    GROUP BY created_date::date
    ORDER BY num_complaints DESC;
SELECT * FROM nyc_311_flooding_daily_complaints LIMIT 10;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
10 rows affected.


[(datetime.date(2021, 9, 2), 645),
 (datetime.date(2011, 8, 25), 623),
 (datetime.date(2014, 12, 9), 537),
 (datetime.date(2021, 9, 1), 530),
 (datetime.date(2017, 5, 5), 492),
 (datetime.date(2021, 10, 26), 472),
 (datetime.date(2014, 4, 30), 446),
 (datetime.date(2013, 5, 8), 406),
 (datetime.date(2010, 3, 30), 381),
 (datetime.date(2011, 8, 26), 379)]

In [52]:

%%sql
CREATE TABLE IF NOT EXISTS nyc_311_flooding_mlt AS
SELECT 
    EXTRACT('MONTH' FROM date) as month, 
    AVG(num_complaints)::int as num_complaints_avg, 
    PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY num_complaints) AS num_complaints_iqr25,
    PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY num_complaints) AS num_complaints_iqr75
FROM nyc_311_flooding_daily_complaints
GROUP BY EXTRACT('MONTH' FROM date)
ORDER BY month;

ALTER TABLE nyc_311_flooding_mlt
    ADD COLUMN IF NOT EXISTS num_complaints_iqr numeric,
    ADD COLUMN IF NOT EXISTS num_complaints_iqrmax numeric;

UPDATE nyc_311_flooding_mlt
    SET num_complaints_iqr = num_complaints_iqr75 - num_complaints_iqr25;

-- 75th quartile + 1.5 x Interquartile Range
UPDATE nyc_311_flooding_mlt
    SET num_complaints_iqrmax = num_complaints_iqr75 + (1.5 * num_complaints_iqr);

SELECT * FROM nyc_311_flooding_mlt;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
Done.
12 rows affected.
12 rows affected.
12 rows affected.


[(Decimal('1'), 17, 7.0, 20.0, Decimal('13'), Decimal('39.5')),
 (Decimal('2'), 23, 8.0, 21.0, Decimal('13'), Decimal('40.5')),
 (Decimal('3'), 24, 9.0, 26.0, Decimal('17'), Decimal('51.5')),
 (Decimal('4'), 25, 10.0, 27.75, Decimal('17.75'), Decimal('54.375')),
 (Decimal('5'), 36, 13.25, 39.75, Decimal('26.5'), Decimal('79.5')),
 (Decimal('6'), 36, 15.0, 40.0, Decimal('25'), Decimal('77.5')),
 (Decimal('7'), 36, 16.0, 42.0, Decimal('26'), Decimal('81')),
 (Decimal('8'), 39, 17.0, 42.5, Decimal('25.5'), Decimal('80.75')),
 (Decimal('9'), 36, 13.0, 38.0, Decimal('25'), Decimal('75.5')),
 (Decimal('10'), 28, 11.0, 30.0, Decimal('19'), Decimal('58.5')),
 (Decimal('11'), 23, 9.0, 23.0, Decimal('14'), Decimal('44')),
 (Decimal('12'), 23, 8.0, 25.0, Decimal('17'), Decimal('50.5'))]

In [53]:

%%sql
CREATE TABLE IF NOT EXISTS nyc_311_flooding_events AS
WITH highs AS (
    SELECT 
    dts.date,
    num_complaints,
    CASE
        WHEN num_complaints > num_complaints_iqr75 THEN 1
        ELSE 0
    END num_complaints_gt75,
    num_complaints - num_complaints_iqr75 as num_complaints_diff75,
    CASE
        WHEN num_complaints > num_complaints_iqrmax THEN 1
        ELSE 0
    END num_complaints_gtmax,
    num_complaints - num_complaints_iqrmax as num_complaints_diffmax
    FROM
    nyc_311_flooding_daily_complaints dts
    INNER JOIN nyc_311_flooding_mlt mlt ON EXTRACT('MONTH' FROM dts.date) = mlt.month
    ORDER BY dts.date
)
SELECT 
    highs.*
FROM highs
WHERE 
    highs.num_complaints_gtmax = 1;

SELECT * FROM nyc_311_flooding_events
ORDER BY num_complaints_diff75 DESC
LIMIT 10;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
10 rows affected.


[(datetime.date(2021, 9, 2), 645, 1, 607.0, 1, Decimal('569.5')),
 (datetime.date(2011, 8, 25), 623, 1, 580.5, 1, Decimal('542.25')),
 (datetime.date(2014, 12, 9), 537, 1, 512.0, 1, Decimal('486.5')),
 (datetime.date(2021, 9, 1), 530, 1, 492.0, 1, Decimal('454.5')),
 (datetime.date(2017, 5, 5), 492, 1, 452.25, 1, Decimal('412.5')),
 (datetime.date(2021, 10, 26), 472, 1, 442.0, 1, Decimal('413.5')),
 (datetime.date(2014, 4, 30), 446, 1, 418.25, 1, Decimal('391.625')),
 (datetime.date(2013, 5, 8), 406, 1, 366.25, 1, Decimal('326.5')),
 (datetime.date(2010, 3, 30), 381, 1, 355.0, 1, Decimal('329.5')),
 (datetime.date(2011, 8, 26), 379, 1, 336.5, 1, Decimal('298.25'))]

In [54]:
%%sql 
SELECT COUNT(*) 
FROM nyc_311_flooding_events
WHERE num_complaints_gtmax = 1;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
1 rows affected.


[(391,)]

Let's add a date column & index to the full record to facilitate quick  joins between our summary table and the full record

In [55]:
%%sql
ALTER TABLE nyc_311_flooding
ADD COLUMN IF NOT EXISTS "date" date,
ADD COLUMN IF NOT EXISTS "month" int;

UPDATE nyc_311_flooding
SET date = created_date::date;

UPDATE nyc_311_flooding
SET month = EXTRACT("MONTH" FROM created_date)::int;

CREATE INDEX IF NOT EXISTS nyc_311_flooding_date_idx ON nyc_311_flooding(date);
CREATE INDEX IF NOT EXISTS nyc_311_flooding_month_idx ON nyc_311_flooding(month);

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
136108 rows affected.
136108 rows affected.
Done.
Done.


[]

## NYC Neighborhood Tabulation Areas (NTAs)

We will use these as aggregation polygons for our analysis.

In [56]:
shp_name = 'geo_export_2296f1c1-ac4f-4cca-b927-e2e39cfcb876'
nta_shp = f'../data/tmp/nyc_ntas_2010/{shp_name}.shp'

In [57]:
%%bash -s "$nta_shp" "$shp_name"
unzip -o ../data/nyc_ntas_2010.zip -d ../data/tmp/nyc_ntas_2010 > /dev/null
gdalsrsinfo $1 | grep "PROJ"

PROJ.4 : +proj=longlat +ellps=WGS84 +no_defs


In [58]:
%%bash -s "$nta_shp" "$shp_name"
ogr2ogr -f "PostgreSQL" \
 PG:"host='0.0.0.0' port='25432' user='docker' password='docker' dbname='storms'" $1 $2 \
 -nlt PROMOTE_TO_MULTI \
 -nln "nyc_nta_2010" \
 -lco GEOMETRY_NAME=geom \
 -lco precision=NO \
 -overwrite 

# cleanup
rm -r ../data/tmp/nyc_ntas_2010


In [59]:
%sql SELECT * FROM nyc_nta_2010 LIMIT 5;

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
5 rows affected.


[(1, 4.0, 'Queens', '081', 'QN08', 'St. Albans', 77412747.7931, 45401.3169214, '010600002032BF0D0001000000010300000001000000D600000084F51C99217052C03B3479E9445A44407BFC9A961C7052C046CA83912E5A4440872F248B177052C06225E539185A44404 ... (6602 characters truncated) ... 052C0BF3B8284265A4440D80F80304D7052C01BB72F4C2E5A4440C0CCD7A63E7052C0166652C4355A4440C24AE821307052C0FA9C96613D5A444084F51C99217052C03B3479E9445A4440'),
 (2, 2.0, 'Bronx', '005', 'BX28', 'Van Cortlandt Village', 25666124.5948, 21945.719299, '010600002032BF0D00010000000103000000010000008A000000FA725773C57852C0BA3EED5D32714440B8206066C57852C09B8129A12E7144407FCA2B50C57852C0DC48F83A28714440C ... (4170 characters truncated) ... 852C018E8A56C37714440D4704492DB7852C0D0F02694347144408EAFDFDBD07852C0B20F305630714440F0FB0936CE7852C098CCC2492F714440FA725773C57852C0BA3EED5D32714440'),
 (7, 2.0, 'Bronx', '005', 'BX41', 'Mount Hope', 14716710.6424, 18937.2474976, '010600002032BF0D000100000001030000000100000055000000D1B8B5A7517952C020C5D68F626D44408D5D974B627952C04D33CD24286D44406197DCDF697952C0F929E74B0D6D4440A ... (2474 characters truncated) ... 952C08BA9E2D24C6D444033A933EE7D7952C06C4634E0516D444044C10C136A7952C0CB051F49766D4440042215485F7952C0C855119E6D6D4440D1B8B5A7517952C020C5D68F626D4440'),
 (3, 4.0, 'Queens', '081', 'QN55', 'South Ozone Park', 82461393.7722, 36708.169488, '010600002032BF0D00010000000103000000010000009A000000A797A1B5917352C04EE61E546A5744408CB4D69B8D7352C01419F9265957444073EC1AD6887352C0BB1A6A2945574440D ... (4682 characters truncated) ... 352C0EAE369F5D2574440D777795FA77352C0D8FBE8BBC25744405B34246F9F7352C0C8AABCA1A3574440C8752F85967352C09BAF94B57E574440A797A1B5917352C04EE61E546A574440'),
 (4, 3.0, 'Brooklyn', '047', 'BK40', 'Windsor Terrace', 14041667.9508, 19033.6718103, '010600002032BF0D000100000001030000000100000086000000C877B521BB7E52C0A7F542B7A0544440E31925B3BB7E52C0E0F5B31E9F5444406DA1D5B3BB7E52C0A029AC1C9F5444407 ... (4042 characters truncated) ... E52C070C19D627F5444405815635BED7E52C04FEDE08E9954444078EBF316E57E52C030E16753AD5444400926A7DCDC7E52C0D46F6C0DC1544440C877B521BB7E52C0A7F542B7A0544440')]

In [60]:
%sql SELECT UpdateGeometrySRID('nyc_nta_2010', 'geom', 4326);

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
1 rows affected.


[('public.nyc_nta_2010.geom SRID changed to 4326',)]

In [61]:
%%sql

ALTER TABLE nyc_nta_2010
ADD COLUMN IF NOT EXISTS geog geography;

UPDATE nyc_nta_2010
SET geog = geom::geography;

ALTER TABLE nyc_nta_2010
ADD COLUMN IF NOT EXISTS geom_2263 geometry;

UPDATE nyc_nta_2010
SET geom_2263 = ST_Transform(geom, 2263);

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
195 rows affected.
Done.
195 rows affected.


[]

In [62]:
%%sql
-- add spaital indexes
CREATE INDEX IF NOT EXISTS nyc_ntas_2010_geom_idx on nyc_nta_2010 USING gist(geom);
CREATE INDEX IF NOT EXISTS nyc_ntas_2010_geom_2263_idx on nyc_nta_2010 USING gist(geom_2263);
CREATE INDEX IF NOT EXISTS nyc_nta_2010_geog_idx on nyc_nta_2010 USING gist(geog);

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
Done.
Done.
Done.


[]

This bit is just for re-running the code. 

In [64]:
%sql DELETE FROM sites WHERE id > 7

 * postgresql+psycopg2://docker:***@0.0.0.0:25432/storms
0 rows affected.


[]